pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -125,13 +125,64 @@ func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
 	// where files will land inside the new image.
 	scriptsDestDir := filepath.Join(getDestination(config), "scripts")
 	sourceDestDir := filepath.Join(getDestination(config), "src")
+	artifactsDestDir := filepath.Join(getDestination(config), "artifacts")
+	artifactsTar := sanitize(filepath.ToSlash(filepath.Join("/tmp", "artifacts.tar")))
+
+	// only COPY scripts dir if required scripts are present, i.e. the dir is not empty;
+	// even if the "scripts" dir exists, the COPY would fail if it was empty
+	scriptsProvided, fileNames := checkValidDirWithContents(filepath.Join(config.WorkingDir, builder.uploadScriptsDir))
+	assembleProvided := false
+	runProvided := false
+	saveArtifactsProvided := false
+	for _, f := range fileNames {
+		glog.V(2).Infof("found override script file %s", f.Name())
+		if f.Name() == "run" {
+			runProvided = true
+		} else if f.Name() == "assemble" {
+			assembleProvided = true
+		} else if f.Name() == "save-artifacts" {
+			saveArtifactsProvided = true
+		}
+		if runProvided && assembleProvided && saveArtifactsProvided {
+			break
+		}
+	}
+
+	if config.Incremental {
+		imageTag := util.FirstNonEmpty(config.IncrementalFromTag, config.Tag)
+		if len(imageTag) == 0 {
+			return errors.New("Image tag is missing for incremental build")
+		}
+		buffer.WriteString(fmt.Sprintf("FROM %s as cached\n", imageTag))
+		if len(config.AssembleUser) > 0 {
+			buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+		}
+		var artifactsScript string
+		if saveArtifactsProvided {
+			glog.V(2).Infof("Override save-artifacts script is included in directory %q", builder.uploadScriptsDir)
+			buffer.WriteString("# Copying in override save-artifacts script\n")
+			artifactsScript = sanitize(filepath.ToSlash(filepath.Join(scriptsDestDir, "save-artifacts")))
+			uploadScript := sanitize(filepath.ToSlash(filepath.Join(builder.uploadScriptsDir, "save-artifacts")))
+			buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), uploadScript, artifactsScript))
+		} else {
+			buffer.WriteString(fmt.Sprintf("# Save-artifacts script sourced from builder image based on user input or image metadata.\n"))
+			buffer.WriteString(fmt.Sprintf("# If this file does not exist in the image, the build will fail.\n"))
+			artifactsScript = sanitize(filepath.ToSlash(filepath.Join(config.ImageScriptsDir, "save-artifacts")))
+		}
+		buffer.WriteString(fmt.Sprintf("RUN %s > %s\n", artifactsScript, artifactsTar))
+	}
 
 	buffer.WriteString(fmt.Sprintf("FROM %s\n", config.BuilderImage))
 
 	if len(config.AssembleUser) > 0 {
 		buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
 	}
 
+	if config.Incremental {
+		buffer.WriteString(fmt.Sprintf("COPY --from=cached --chown=%s:0 %s %s\n", sanitize(imageUser), artifactsTar, artifactsTar))
```

Comment:
Are `--from` and `--chown` newer dockerfile options?  Some albeit quick internet searches were not conclusive for me ...

Ultimately, while --as-dockefile is experimental, the docker version noted at https://github.com/openshift/source-to-image#dependencies seems very very old ...

Not directly specific to this PR, but to --as-dockerfile in general:  do we need to button up the docker dependencies any?

Reply:
> Are --from and --chown newer dockerfile options? Some albeit quick internet searches were not conclusive for me ...

yes. newish anyway.  they aren't in the fedora docker version.

> Not directly specific to this PR, but to --as-dockerfile in general: do we need to button up the docker dependencies any?

Since we say "1.6 or greater", i would say not unless we think it no longer works w/ the old versions we indicate for some reason.

Note that s2i itself, when used with --as-dockerfile, has no docker dependency at all.  The dockerfile it generates, of course, will only work with a new enough docker version (or buildah, or some other image building tool that consumes dockerfiles and understands the newer syntax options)

Reply:
> Is there any standardization around what a Dockerfile supports? 

:rofl: